src/libraries/System.Private.CoreLib/src/System/Text/ASCIIUtility.cs

```diff
@@ -1435,18 +1768,46 @@ private static unsafe nuint NarrowUtf16ToAscii_Sse2(char* pUtf16Buffer, byte* pA
             }
             else
             {
-                if ((Sse2.MoveMask(Sse2.AddSaturate(utf16VectorFirst.AsUInt16(), asciiMaskForAddSaturate).AsByte()) & NonAsciiDataSeenMask) != 0)
+                if (Sse2.IsSupported)
                 {
-                    goto Finish; // found non-ASCII data
+                    if ((Sse2.MoveMask(Sse2.AddSaturate(utf16VectorFirst.AsUInt16(), asciiMaskForAddSaturate).AsByte()) & NonAsciiDataSeenMask) != 0)
+                    {
+                        goto Finish; // found non-ASCII data
+                    }
+                }
+                else if (AdvSimd.Arm64.IsSupported)
+                {
+                    Debug.Assert(AdvSimd.IsSupported);
+                    if (ContainsNonAsciiByte(AdvSimd.AddSaturate(utf16VectorFirst.AsUInt16(), asciiMaskForAddSaturate)))
+                    {
+                        goto Finish;
+                    }
+                }
+                else
+                {
+                    throw new PlatformNotSupportedException();
                 }
             }
 
             // First part was all ASCII, narrow and aligned write. Note we're only filling in the low half of the vector.
-            asciiVector = Sse2.PackUnsignedSaturate(utf16VectorFirst, utf16VectorFirst);
-
-            Debug.Assert(((nuint)pAsciiBuffer + currentOffsetInElements) % sizeof(ulong) == 0, "Destination should be ulong-aligned.");
+            if (Sse2.IsSupported)
+            {
+                asciiVector = Sse2.PackUnsignedSaturate(utf16VectorFirst, utf16VectorFirst);
+                Debug.Assert(((nuint)pAsciiBuffer + currentOffsetInElements) % sizeof(ulong) == 0, "Destination should be ulong-aligned.");
+                Sse2.StoreScalar((ulong*)(pAsciiBuffer + currentOffsetInElements), asciiVector.AsUInt64()); // ulong* calculated here is aligned
+            }
+            else if (AdvSimd.Arm64.IsSupported)
+            {
+                Vector64<byte> lower = AdvSimd.ExtractNarrowingSaturateUnsignedLower(utf16VectorFirst);
+                asciiVector = AdvSimd.ExtractNarrowingSaturateUnsignedUpper(lower, utf16VectorFirst).AsByte();
+                Debug.Assert(((nuint)pAsciiBuffer + currentOffsetInElements) % sizeof(ulong) == 0, "Destination should be ulong-aligned.");
```

Comment:
>Debug.Assert(((nuint)pAsciiBuffer + currentOffsetInElements) % sizeof(ulong) == 0, "Destination should be ulong-aligned."); [](start = 16, length = 123)

should this assert and the one inside `SSE2` outside and should hold true regardless?

Reply:
or may be not. Ignore.

---
In reply to: [454718698](https://github.com/dotnet/runtime/pull/39103#discussion_r454718698) [](ancestors = 454718698)